### ISSUE TITLE
Project proposal: light-4j microservices framework

### ISSUE THREAD

**Issue by stevehu** [cncf/toc#79]
As suggested by @caniszczyk, we submit light-4j for consideration to be included in CNCF projects.

Github repo: https://github.com/networknt/light-4j
Document: https://doc.networknt.com/

More information can be be found in the following document. 

https://docs.google.com/document/d/1p1v2Fy05SUy5vTTPMhySdpT7KsHJNh4-u7QgGdKEOAA/edit?usp=sharing

To get started with RESTful APIs, these two tutorials will be a good starting points. 

https://doc.networknt.com/tutorial/rest/openapi/petstore/
https://doc.networknt.com/tutorial/rest/swagger/petstore/

Our own benchmarks
https://github.com/networknt/microservices-framework-benchmark

Third party benchmarks
https://www.techempower.com/benchmarks/previews/round15/

**Comment by adrianmaurer** [comment]
looks like you just glued a bunch of libraries and utilities together on top of undertow? What if I don't want all that stuff and compose something that works for my specific needs?

**Comment by stevehu** [comment]
@adrianmaurer Yes. We basically built dozens of middleware handlers that address all cross-cutting concerns for most microservices running in a container. In addition, we provide handlers to support different style of APIs like Swagger 2.0, OpenAPI 3.0, GraphQL and RPC. If users want to build services with asynchronous events, we have Event Sourcing/CQRS, transactional messaging and saga frameworks to choose from. 

All the handlers are glued together with service module so that you can add or remove with an externalized config file. Also, every handler has its own configuration file to manipulate behaviors. The light-codegen is used to scaffold a new project with default handlers wired in; however, users can add other provided handlers or their in-house built handlers based on the application requirement. 

We also provide light-oauth2, light-portal, tokenization services to help manage life cycles of microservices at scale. 

You can use Undertow to build your service but light platform provides an ecosystem for more development productivity. With the scaffolded project, developers can only focus on the business handlers and test cases for these handlers without worrying about all other cross-cutting concerns. 

With all the cross-cutting concerns built-in, every service built on top of our frameworks has an embedded gateway so we don't need any API gateway or Service Mesh in the deployment.

**Comment by adrianmaurer** [comment]
Still sounds to me like light4j brings on a lot of assumptions and baggage. I think I understand what you're trying to do and that is to provide a reference architecture and some standard dev tools. But I'm not sure light4j achieves that. And here is why

1. Everything you've selected to be part of the ecosystem are more or less wrappers around common tools and patterns with opinions so that they integrate with light4j. But what if I don't have a need to apply those patterns or use those tools. For example, if I'm already using graphite, do I need to tear that down and migrate to prometheus now?

2. How light4j has selected tools make it very difficult for an application developer to evolve their app over time.  There are many alternatives to the selected tools and better ones may be released over time.  I'd prefer to include the tool directly and build my abstractions around it myself.  How to create an manage my infrastructure is also tightly coupled creating more dependences. 

3. Light4j presumes that I need a bunch of patterns in my application, like audit for example. What gives light4j the right to tell me how I apply specific patterns to my application?

And fine, maybe if I was trying to setup a standard on building applications at a large org I may want to build a reference implementation and select some standard dev tools around some best practices.  But that should be a decision made within the organization not a framework developer. And even then, a reference architecture is just that, a reference on how with some opinions vs "thou shall". Especially as large organizations move towards more product oriented development where product teams have greater autonomy on how to build their applications -- team chooses their best practices, patterns, tools and how to apply them.

And you can use tools like yeoman to do your initial scaffolding based on the latest reference architecture (which you may have or need many).

If I were to sum it up, Java already has a large ecosystem of libraries and tools to select from. Dependency management is already very mature. Why would I select an opinionated abstraction on top of that ecosystem that keeps me tightly bound to it. Isn't that an anti-pattern? Especially for something open source? 

In other words, as I mentioned in my first comment, light4j has just glued a bunch of current libraries together and a very opinionated manner. In 2-3 years from now, without a huge renovation, using light4j will seem like buying a used home.

**Comment by stevehu** [comment]
We are not trying to provide a reference architecture but trying to provide a skeleton that allows users to create their own handlers to attach/inject to it. It is just like we build a Lego base and provide some blocks as references. Other players can use our blocks or build something totally different. 

Although light-4j is built on top of Undertow, it has the same performance as you build the application manually. We call it a zero-cost abstraction. You can see the performance comparison at https://www.techempower.com/benchmarks/

For the metrics collection example, we have provided hander for [InfluxDB](https://www.networknt.com/concern/metrics/) and [Prometheus](https://www.networknt.com/concern/prometheus/) as these are typical push and pull implementations. Other users can create their own handlers and replace the reference metrics handler depending on their environment and non-functional requirement. 

As I have explained above, we don't select tools for our users but only provide some reference implementations so that user can learn how to leverage their tools in the framework. At the core of the framework, we do have some dependencies and we are trying out best to select green libraries without too many dependencies. This can make the core of the framework very small and reduce the risk of dependency conflicts. 

I don't think there is a fixed pattern or somehow the framework is forcing the user to choose one or the other. We have provided two built-in audit handlers: one to dump only the important info and the other dump everything in request/response into audit log file. One of the bank customers created an audit handler to save information into Oracle database. 

 I totally agree with you that each organization should have the flexibility to choose best practices, patterns and tools. I am actually collecting these best practices from my customers [here](https://www.networknt.com/development/best-practices/).

We have our own generator light-codegen which is very similar to yeoman.

The dependency injection is opinionated because we want to have externalized config files and only want to support constructor injection to avoid confusion. Most injections are happening in the startup hooks anyway. 

For Java EE developers, it is very hard to understand the modern async nature of the Undertow, we are just trying to provide an abstract layer to hide these cross-cutting concerns so that developers can only focus on the business logic.

**Comment by caniszczyk** [comment]
closing, no interest from @cncf/toc to present
